crates/nu-command/src/filesystem/ls.rs

```diff
@@ -338,92 +366,129 @@ fn ls_for_one_pattern(
         });
     }
 
-    let mut hidden_dirs = vec![];
+    let hidden_dirs = Arc::new(Mutex::new(Vec::new()));
 
-    let signals = signals.clone();
-    Ok(Box::new(paths_peek.filter_map(move |x| match x {
-        Ok(path) => {
-            let metadata = match std::fs::symlink_metadata(&path) {
-                Ok(metadata) => Some(metadata),
-                Err(_) => None,
-            };
-            if path_contains_hidden_folder(&path, &hidden_dirs) {
-                return None;
-            }
+    let signals_clone = signals.clone();
 
-            if !all && !hidden_dir_specified && is_hidden_dir(&path) {
-                if path.is_dir() {
-                    hidden_dirs.push(path);
-                }
-                return None;
-            }
+    let pool = if use_threads {
+        let count = std::thread::available_parallelism()?.get();
+        create_pool(count)?
+    } else {
+        create_pool(1)?
+    };
+
+    pool.install(|| {
+        paths_peek
+            .par_bridge()
+            .filter_map(move |x| match x {
+                Ok(path) => {
+                    let metadata = match std::fs::symlink_metadata(&path) {
+                        Ok(metadata) => Some(metadata),
+                        Err(_) => None,
+                    };
+                    let hidden_dir_clone = Arc::clone(&hidden_dirs);
+                    let mut hidden_dir_mutex = hidden_dir_clone
+                        .lock()
+                        .expect("Unable to acquire lock for hidden_dirs");
+                    if path_contains_hidden_folder(&path, &hidden_dir_mutex) {
+                        return None;
+                    }
+
+                    if !all && !hidden_dir_specified && is_hidden_dir(&path) {
+                        if path.is_dir() {
+                            hidden_dir_mutex.push(path);
+                        }
+                        return None;
+                    }
```

Comment:
I think the hidden dir mutex is still locked, you might want to drop it or use some other kind of strategy (maybe parallelize looking up the metadata but serialize from there)